### DISCUSSION TITLE
Revolve

### DISCUSSION THREAD

**Discussion by jgomez720** [KittyCAD/modeling-app#496] [New Feature Requests]
### Description
A tool that enables the user to revolve a sketch around an axis.

### Functionality
1. The user selects a sketch profile.
2. The user selects an axis to revolve around. This can be the coordinate system axes, an edge on a body, and in more advanced cases, the centerline of any cylindrical cutout.
3. The user specifies the angle in which to revolve (meaning it doesn't have to go all the way around, 360 degrees)

#### Full Rotation
<img width="364" alt="Screenshot 2023-11-06 at 11 19 32 AM" src="https://github.com/KittyCAD/modeling-app/assets/114548659/8a3fc5e4-9bfb-47a5-ac4f-b0c978559712">
<img width="699" alt="Screenshot 2023-11-06 at 11 20 06 AM" src="https://github.com/KittyCAD/modeling-app/assets/114548659/19dbfd07-223c-4459-9736-b31085a91024">
<img width="456" alt="Screenshot 2023-11-06 at 11 20 14 AM" src="https://github.com/KittyCAD/modeling-app/assets/114548659/551e8b16-92d1-4cb6-95ce-7f8ce2a9fea6">

#### Partial Rotation
<img width="692" alt="Screenshot 2023-11-06 at 11 21 29 AM" src="https://github.com/KittyCAD/modeling-app/assets/114548659/a1f20a62-43eb-4e2c-b224-f2b63a6ee6c3">
<img width="482" alt="Screenshot 2023-11-06 at 11 21 35 AM" src="https://github.com/KittyCAD/modeling-app/assets/114548659/11bf3ba6-308e-4281-aef9-dc9de8cde706">

## PseudoKCL
```
const part001 = startSketchOn('XY')
  |> startProfileAt([4, 12], %)
  |> line([2, 0], %)
  |> line([0, -6], %)
  |> line([4, -6], %)
  |> line([0, -6], %)
  |> line([-3.75, -4.5], %)
  |> line([0, -5.5], %)
  |> line([-2, 0], %)
  |> close(%)

revolve({
    sketch: part001,
    angle: 180, //optional, default to 360
    axis: 'absoluteY' // absolute meaning world coordinate system
})
```
The sketch would look like this:
<img width="573" alt="Screenshot 2023-11-06 at 10 15 13 AM" src="https://github.com/KittyCAD/modeling-app/assets/114548659/408a12c2-4b05-47a9-a7d6-2d29e9769f2c">

**End Result:**
<img width="425" alt="Screenshot 2023-11-06 at 10 12 24 AM" src="https://github.com/KittyCAD/modeling-app/assets/114548659/5bac0f2d-a39e-4ce5-adf4-25f6d6ee1c00">

OR to revolve around an axis the user specifies
```
const part001 = startSketchOn('XY')
  |> startProfileAt([4, 12], %)
  |> line([2, 0], %)
  |> line([0, -6], %)
  |> line([4, -6], %)
  |> line([0, -6], %)
  |> line([-3.75, -4.5], %)
  |> line([0, -5.5], %)
  |> line([-2, 0], %)
  |> close(%)
  |> lineSingle([0,0],[0, 5], construction: true, tag: 'revolveAxis' %)

revolve({
    sketch: part001,
    angle: 180, //optional, default to 360
    axis: getEdge(part001, 'revolveAxis') // User-specified axis
})
```

## Advanced KCL
```
const box = startSketchOn('XY')
  |> startProfileAt([0, 0], %)
  |> line([0, 5], %)
  |> line([5, 0], %)
  |> line([0, -5], %, tag: 'revolveAxis') // tag = sketch line, endTag = the line created after an extrude
  |> close(%)
  |> extrude(5, %)

const sketch001 = startSketchOn(extrusionEnd(box)) // meaning the face created by an extrude. The 'end' of the extrude
  |> circle([7, 2.5], 1, %)

revolve({
    sketch: sketch001,
    angle: 90, //optional, default to 360
    axis: getEdge(box, 'revolveAxis', end: true) // absolute meaning world coordinate system
})
```
**box and circle sketch**
<img width="548" alt="Screenshot 2023-11-06 at 10 18 44 AM" src="https://github.com/KittyCAD/modeling-app/assets/114548659/a68383f5-826a-4ffe-8b31-6e991358e0f6">

**Revolve**
<img width="571" alt="Screenshot 2023-11-06 at 10 18 37 AM" src="https://github.com/KittyCAD/modeling-app/assets/114548659/0389bede-779a-4c65-b6fa-52ae4b3be9af">

**Comment by mlfarrell** [comment]
engine-side support for this feature is now done
